hardware/detect.py

```diff
@@ -689,25 +689,105 @@ def find_element(xml, xml_spec, sys_subtype,
     return True
 
 
-def get_cpus(hw_lst):
-    # Extracting lspcu information
-    lscpu = {}
-    output = detect_utils.output_lines('LANG=en_US.UTF-8 lscpu')
+def _get_cpufreq_govenors(hw_lst, nr_cpus):
+    # Govenors could be differents on logical cpus
+    for cpu in range(nr_cpus):
+        scaling_governor = "/sys/devices/system/cpu/cpufreq/policy{}/scaling_governor".format(
+            cpu)
+        if os.path.exists(scaling_governor):
+            with open(scaling_governor) as governorfile:
+                hw_lst.append(('cpu', "logical_{}".format(cpu),
+                               "governor", governorfile.readline().rstrip('\n')))
 
-    for line in output:
-        if ':' in line:
-            item, value = line.split(':')
-            lscpu[item.strip(':')] = value.strip()
 
-    # Extracting lspcu -x information
-    lscpux = {}
-    output = detect_utils.output_lines('LANG=en_US.UTF-8 lscpu -x')
+def _get_cpu_numa_layout(hw_lst, lscpu, lscpux):
+    # Extracting numa nodes
+    hw_lst.append(('numa', 'nodes', 'count', int(lscpu['NUMA node(s)'])))
+    # ppc64le has sparse NUMA nodes eg:
+    #   NUMA node0 CPU(s):   ffffffffffffffffff
+    #   NUMA node8 CPU(s):   ffffffffffffffffff000000000000000000
+    #   NUMA node252 CPU(s): 0
+    #   NUMA node253 CPU(s): 0
+    #   NUMA node254 CPU(s): 0
+    #   NUMA node255 CPU(s): 0
+    # So get the keys in order
+
+    numa_nodes = [(k, int(k[9:-7]))
+                  for k in lscpux.keys()
+                  if 'NUMA node' in k and 'CPU(s)' in k]
+    for (key, node_id) in numa_nodes:
+        cpus = lscpu[key]
+        # lscpu -x provides the cpu mask
+        cpu_mask = lscpux[key]
+        total_cpus = 0
+        min_cpu = None
+        max_cpu = None
 
-    for line in output:
-        if ':' in line:
-            item, value = line.split(':')
-            lscpux[item.strip(':')] = value.strip()
+        # ppc64le often has NUMA nodes without online CPUs
+        if cpus:
+            for item in cpus.split(','):
+                # lscpu report numa nodes like 0-5,48-53
+                if "-" in item:
+                    max_cpu = int(item.split("-")[1])
+                    min_cpu = int(item.split("-")[0])
+                    total_cpus = total_cpus + max_cpu - min_cpu + 1
+                else:
+                    # or like 0,1
+                    if min_cpu is None:
+                        min_cpu = int(item)
+                    else:
+                        max_cpu = int(item)
+                        total_cpus = total_cpus + max_cpu - min_cpu + 1
+
+        # total_cpus = 12 for "0-5,48-53"
+        hw_lst.append(('numa', 'node_{}'.format(
+            node_id), 'cpu_count', total_cpus))
+        hw_lst.append(('numa', 'node_{}'.format(
+            node_id), 'cpu_mask', cpu_mask))
+
+
+def _ppc64le_cpu_info(hw_lst, lscpu, lscpux):
```

Comment:
This call is mostly identical to the x86_64 one, can we use one function and just check for keys presence?

Reply:
Yup I'm working on that approach it's pretty messy though so I;m trying a few options